perf/perf_json.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+#
+# Copyright: 2024 IBM
+# Author: Disha Goel <disgoel@linux.ibm.com>
+
+import os
+import platform
+import json
+import shutil
+from avocado import Test
+from avocado.utils import cpu, distro, dmesg, process
+from avocado.utils.software_manager.manager import SoftwareManager
+
+
+class perf_json(Test):
+
+    """
+    This tests all pmu events
+    :avocado: tags=perf,json,events
+    """
+    testdir = 'tools/perf/pmu-events/arch/powerpc/'
+
+    # Initializing fail command list
+    fail_cmd = list()
+
+    def setUp(self):
+        """
+        Setup checks:
+        0. Processor should be ppc64.
+        1. Install perf package
+        2. Install and build kernel source
+        3. Collect all the pmu events from perf list
+        4. Run all the pmu events using perf stat
+        5. Collect all the events and event code from json files
+        6. Compare both the events list and event code
+        """
+        smm = SoftwareManager()
+
+        detected_distro = distro.detect()
+        if 'ppc64' not in detected_distro.arch:
+            self.cancel("Processor is not PowerPC")
+
+        # make sure kernel source repo is configured
+        deps = ['gcc', 'make', 'perf']
```

Comment:
Can you also add upstream as a target so that the test can fetch upstream kernel source? This will allow the test to be executed against a host booted with upstream kernel.

Reply:
I have added upstream target also and pushed code.